src/components/FileTree/index.tsx

```diff
@@ -184,4 +192,17 @@ export class FileTreeBase extends React.Component<Props> {
   }
 }
 
-export default connect()(FileTreeBase) as React.ComponentType<PublicProps>;
+const mapStateToProps = (
+  state: ApplicationState,
+  ownProps: PublicProps,
+): PropsFromState => {
+  const { versionId } = ownProps;
+
+  return {
+    version: getVersionInfo(state.versions, versionId),
```

Comment:
The return type of `getVersionInfo` [is wrong](https://github.com/mozilla/addons-code-manager/issues/442) so your component needs to account for the fact that it can be `void` (when not loaded).

Introducing the logic of fetching `version` from state means that `FileTree` *should* be responsible for loading that version into state when needed. This would be the only way to achieve true encapsulation and avoid confusion in future refactoring. In your current patch, `FileTree` is tightly coupled to the components that know how to load versions, which are [`Browse`](https://github.com/mozilla/addons-code-manager/blob/ee94f3282cd30df66a634f49bda895121fedb8a4/src/pages/Browse/index.tsx#L58) or [`Compare`](https://github.com/mozilla/addons-code-manager/blob/ee94f3282cd30df66a634f49bda895121fedb8a4/src/pages/Compare/index.tsx#L87). Before your patch, when `version` was passed as a direct prop, the coupling was obvious. Now it is no longer obvious and that can lead to future bugs and confusion.

So, what to do? The simplest thing is to keep passing in `version` because it makes the coupling obvious. I suggested not doing this because it made the story code more complicated.

The next options are:

1. Introduce loading code in `componentDidMount` for when `version` doesn't exist in state. This loading code would probably never get executed in real life but at least it would be there for when we need it.

2. We probably won't ever need the loading code and loading code adds complexity so I think it would be fine to log a warning to the console when `version` does not exist in state. This log message would help in debugging a story, test, or a future refactoring.

This was a long way of saying "just add a log message" but I wanted to explain why it is a problem and why we should strive to avoid this kind of pattern whenever possible.

Also, when we switch to hooks, we could probably make a shared hook based on [`useEffect`](https://reactjs.org/docs/hooks-reference.html#useeffect) like `useVersionLoader` and fix it The Right Way ™.

Reply:
This is all good information, thanks @kumar303. It does sound like the `FileTree` should be responsible for fetching its own `Version` if needed, but, as you report, we know that the `Browse` and `Compare` components will have already done that, so with the current set up we'll never need to do it.

I'll go ahead and add the warning when a version does not exist.

Reply:
Ok, I've added the warning. Do you think we need a test for the warning as well, or are we happy to just have the code there, as we never expect it to be triggered anyway?

Also, should I change the type of the `version` prop to `Version | void`? I can do that, but then it also requires some extra code in the component to deal with the case where `version` is `null`. For example, we'd need to deal with it in `renderNode`.